spec/u-namespace.md

```diff
@@ -0,0 +1,72 @@
+# MessageFormat 2.0 Unicode Namespace
+
+The `u:` namespace is reserved for use by the Unicode Consortium.
+
+## Options
+
+This section describes common **_<dfn>`u:` options</dfn>_** which each implementation SHOULD support
+for all _functions_ and _markup_.
+
+### `u:id`
+
+A string value that is included as an `id` or other suitable value
+in the formatted parts for the _placeholder_,
+or any other structured formatted results.
+
+Ignored when formatting a message to a string.
+
+Accepts string values, or values which can be stringified without error.
+For other values, a _Bad Option_ error is emitted
+and the `u:id` option is ignored.
+
+### `u:locale`
+
+A comma-delimited list consisting of
+well-formed [BCP 47](https://www.rfc-editor.org/rfc/bcp/bcp47.txt)
+language tags,
+or an implementation-defined list of such tags.
+
+Replaces the _locale_ defined in the _function context_ for this _expression_.
+
+If this option is set on _markup_, a _Bad Option_ error is emitted
+and the value of the `u:locale` option is ignored.
+
+During processing, the `u:locale` option
+MUST be removed from the resolved mapping of _options_
+before calling the _function handler_.
+
+Values matching the following ABNF are always accepted:
+```abnf
+u-locale-option = unicode_bcp47_locale_id *(o "," o unicode_bcp47_locale_id)
+```
+using `unicode_bcp47_locale_id` as defined for
+[Unicode Locale Identifier](https://cldr-smoke.unicode.org/spec/main/ldml/tr35.html#unicode_bcp47_locale_id).
+
+Implementations MAY support additional language tags,
+such as private-use or grandfathered tags,
+or tags using `_` instead of `-` as a separator.
+When the value of `u:locale` is set by a _variable_,
+implementations MAY support non-string values otherwise representing locales.
+
+For unsupported values, a _Bad Option_ error is emitted
+and the value of the `u:locale` option is ignored.
```

Comment:
Unsupported is unclear here. We also need to resolve list handling.

```
{$foo :date u:locale=|en-US, tlh-Xsux-AQ, i-klingon, bad-Juju, American|}`
```

Does the above work because en-US is fine? Or emit Bad Option because:
- `tlh-Xsux-AQ` a valid tag for Klingon as written in cuneiform as used in Antarctica is not an installed locale
- `i-klingon` is grandfathered and this implementation doesn't support grandfathered tags
- `bad-Juju` is not a valid tag
- `American` is not well-formed

```suggestion
If any of the language tags in the value is unsupported,
invalid, or not well-formed, the implementation ignores that entry in the
list and emits a _Bad Option_ error.
If the resulting list is empty, a _Bad Option_ error is emitted
and the value of the `u:locale` option is ignored.
```

Reply:
I fear that defining this too strictly mandate more than we'd like about how an implementation handles locales. Thinking about this some more, my preference would be to allow for each implementation to follow its own customs on locale fallback chain handling.

For instance JavaScript errors on not well-formed tags, but ignores invalid and unsupported ones:
```js
Intl.NumberFormat.supportedLocalesOf(['en-US', 'tlh-Xsux-AQ', 'i-klingon', 'bad-Juju'])
// Uncaught RangeError: invalid language tag: "i-klingon"

Intl.NumberFormat.supportedLocalesOf(['en-US', 'tok', 'bad-Juju', 'American'])
// [ "en-US" ]
```
but it's not reasonable to expect all other implementations to do the same thing.

Furthermore, referring to validity should probably be avoided as well, because in this context that would mean needing to carry around data that might not have any other use.

How about this?

```suggestion
Implementations MAY emit a _Bad Option_ error
and MAY ignore the value of the `u:locale` option
if it contains a locale that it does not support or
a language tag that is not well-formed.
```

Reply:
Mt examples were meant as rhetorical. Your suggestion is fine, except that I think we should allow for the list to be processed as I suggested. I intended that as MAY... basically the implementation can process the list as it likes. But if the list has one bad tag in it, should we ignore the whole thing? That's how it reads now.

Reply:
The current language is intended to allow for an implementation to ignore the whole value and emit an error if it contains a bad tag, or to not do so, or to decide on a case-by-case basis. Note that we also do not specify how and when locale negotiation is meant to happen; that could be handled by the implementation, but it could also be done separately in each function handler.

Reply:
(Back at a real computer)

Perhaps:

```suggestion
Implementations MAY emit a _Bad Option_ error
and MAY ignore the value of the `u:locale` _option_ as a whole
or any of the entries in the list of language tags.
This might be because the locale specified is not supported
or because the language tag is not well-formed,
not valid, or some other reason.
```

Reply:
Eh, seems a bit verbose but works.